akka-http/src/main/java/akka/http/javadsl/server/HttpApp.java

```diff
@@ -0,0 +1,150 @@
+import akka.http.javadsl.settings.ServerSettings;
+import akka.stream.ActorMaterializer;
+import scala.runtime.BoxedUnit;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * API MAY CHANGE - EXPERIMENTAL
+ * Bootstrap trait. It helps booting up an akka-http by only defining the desired routes.
+ * It offers additional hooks to modify the default behavior
+ */
+// @akka.annotation.ApiMayChange // FIXME replace with real annotation once Akka dependency bumped
+public abstract class HttpApp extends AllDirectives {
+
+  private AtomicReference<ServerBinding> serverBinding = new AtomicReference<ServerBinding>();
+
+  /**
+   * Start a server on the specified host and port, using provided settings
+   * Note that this method is blocking
+   */
+  public void startServer(String host, int port, ServerSettings settings) throws ExecutionException, InterruptedException {
+    startServer(host, port, settings, Optional.empty());
+  }
+
+  /**
+   * Start a server on the specified host and port, using provided settings and [[ActorSystem]]
+   * Note that this method is blocking
+   */
+  public void startServer(String host, int port, ServerSettings settings, ActorSystem system) throws ExecutionException, InterruptedException {
+    startServer(host, port, settings, Optional.of(system));
+  }
+
+  /**
+   * Start a server on the specified host and port, using provided settings and [[ActorSystem]] if present
+   * Note that this method is blocking
+   */
+  public void startServer(String host, int port, ServerSettings settings, Optional<ActorSystem> system) throws ExecutionException, InterruptedException {
+
+    final ActorSystem theSystem = system.orElseGet(() -> ActorSystem.create(Logging.simpleName(this).replaceAll("\\$", "")));
+    final ActorMaterializer materializer = ActorMaterializer.create(theSystem);
+
+    CompletionStage<ServerBinding> bindingFuture = Http
+      .get(theSystem)
+      .bindAndHandle(route().flow(theSystem, materializer),
+        ConnectHttp.toHost(host, port),
+        settings,
+        theSystem.log(),
+        materializer);
+
+    bindingFuture.handle((binding, exception) -> {
+      if (exception != null) {
+        postHttpBindingFailure(exception);
+      } else {
+        //setting the server binding for possible future uses in the client
+        serverBinding.set(binding);
+        postHttpBinding(binding);
+      }
+      return null;
+    });
+
+    try {
+      bindingFuture
+        .thenCompose(ignore -> waitForShutdownSignal(theSystem)) // chaining both futures to fail fast
+        .toCompletableFuture()
+        .exceptionally(ignored -> BoxedUnit.UNIT) // If the future fails, we want to complete normally
+        .get(); // It's waiting forever because maybe there is never a shutdown signal
```

Comment:
Regarding using `get` without a timeout, what if the server should run forever. The timeout we specify here will limit the longest time to live for this server.